### PR TITLE
(abandoned) [compiler][rewrite] PropagateScopeDeps hir rewrite

### PR DESCRIPTION
**Pull request by mofeiZ**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #30079

#### Rvalues / temporaries:
In the compiler, unnamed temporaries that represents the evaluation of an expression
In the code snippet below, $1, $2, $3, and $4 are temporaries.
```js
// input
function Component({ bar} ) {
  const x = {a: foo(bar), b: {}};
}
// gets lowered to
[1] $2 = LoadGlobal(global) foo
[2] $3 = LoadLocal bar$1
[3] $4 = Call $2(<unknown> $3)
[4] $5 = Object {  }
[5] $6 = Object { a: $4, b: $5 }
[6] $8 = StoreLocal Const x$7 = $6
```
The compiler currently treats temporaries and named variables (e.g. `x`) differently in this pass.
- named variables may be reassigned (in fact, since we're running after LeaveSSA, a single named identifier's IdentifierId may map to multiple `Identifier` instances -- each with its own scope and mutable range)
- temporaries are replaced with their represented expressions during codegen. This is correct (mostly correct, see #29878) as we're careful to always lower the correct evaluation semantics. However, since we rewrite reactive scopes entirely (to if/else blocks), we need to track temporaries that a scope produces in `ReactiveScope.declarations` and later promote them to named variables.
In the same example, $4, $5, and $6 need to be promoted: $2 ->`t0`,  $5 ->`t1`, and $6 ->`t2`.
```js
[1] $2 = LoadGlobal(global) foo
[2] $3 = LoadLocal bar$1
scope 0:
  [3] $4 = Call $2(<unknown> $3)
scope 1:
  [4] $5 = Object {  }
scope 2:
  [5] $6 = Object { a: $4, b: $5 }
[6] $8 = StoreLocal Const x$7 = $6
```

#### Dependencies
`ReactiveScope.dependencies` records the set of (read-only) values that a reactive scope is dependent on. This is currently limited to just variables (named variables from source and promoted temporaries) and property-loads.
All dependencies we record need to be hoistable -- i.e. reordered to just before the ReactiveScope begins. Not all PropertyLoads are hoistable.

In this example, we should not evaluate `obj.a.b` without before creating x and checking `objIsNull`.
```js
// reduce-reactive-deps/no-uncond.js
function useFoo({ obj, objIsNull }) {
  const x = [];
  if (isFalse(objIsNull)) {
    x.push(obj.a.b);
  }
  return x;
}
```

While other memoization strategies with different constraints exist, the current compiler requires that `ReactiveScope.dependencies` be re-orderable to the beginning of the reactive scope. But.. `PropertyLoad`s from null values will throw `TypeError`. This means that evaluating hoisted dependencies should throw if and only if the source program throws. (It is also a bug if source throws and compiler output does not throw. See https://github.com/facebook/react-forget/pull/2709)

---
### Rough high level overview
1. Pass 1
Walk over instructions to gather every temporary used outside of its defining scope (same as ReactiveFunction version). These determine the sidemaps we produce, as temporaries used outside of their declaring scopes get promoted to named variables later (and are not considered hoistable rvals).
2. Pass 2 (collectHoistablePropertyLoads)
Walk over instructions to generate sidemaps:
  a. temporary identifier -> named variable and property path (e.g. `$3 -> {obj: props, path: ["a", "b"]}`)
  b. block -> accessed variables and properties (e.g. `bb0 -> [ {obj: props, path: ["a", "b"]} ]`)
  c. Walk over control flow graph to understand the set of object and property paths that can be read by each basic block. This analysis:
    - relies on post-dominator trees
    - traverses the CFG from entry (producing the set of variables/paths unconditionally evaluated *before* a block).
    - traverses the CFG from exit (producing the set of variables/paths unconditionally evaluated *after* a block).
4. Pass 3: (collectDependencies)
Walks over instructions again to record dependencies and declarations, using the previously produced sidemaps

Will add more fixture tests (although most cases should be covered in `reduce-reactive-deps`).
Tested by syncing internally and checking compilation output differences ([internal link](https://fburl.com/wiki_markdown/nazsiszd))

---
### Followups:
1. Rewrite function expression deps
This change produces much more optimal output as the compiler now uses the function CFG to understand which variables / paths are assumed to be non-null. However, it may exacerbate [this function-expr hoisting bug](https://github.com/facebook/react/blob/main/compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-invalid-hoisting-functionexpr.ts). A short term fix here is to simply call some form of `collectNonNullObjects` on every function expression to find hoistable variable / paths. In the longer term, we should refactor out `FunctionExpression.deps`.

2. Enable optional paths
(a) don't count optional load temporaries as dependencies (e.g. `collectOptionalLoadRValues(...)`).
(b) add optional paths back. This is a bit tricky as we'll want to implement some merging logic for `ConditionalAccess | OptionalChain | UnconditionalAccess`. In addition, our current optional chain lowering is slightly incorrect / imprecise

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/functionexpr-conditional-access-2.tsx [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// @enableTreatFunctionDepsAsConditional
+import { Stringify } from "shared-runtime";
+
+function Component({ props }) {
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/functionexpr-conditional-access-2.tsx:4]: nit: it's confusing to destructure props and name the identifier `props`, pick a different name?

File: compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+      blockId,
+      scope: scopeStartBlocks.get(blockId) ?? null,
+      preds: block.preds,
+      assumedNonNullObjects,
+    });
+  }
+  return nodes;
+}
+
+function deriveNonNull(fn: HIRFunction, nodes: Map<BlockId, BlockInfo>): void {
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:284]: The bookkeeping/traversal logic is making it really hard to follow the core logic of this pass. I would have expected that a non-null object analysis could use the dominator tree rather than the walking forward/backwards that's happening here. Did you try using dominators? If yeah, what was the challenge?

Comment (mofeiZ) [reply compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:284]: From my understanding (please correct if I'm misunderstanding something, my algo knowledge is much weaker than yours), dominator tree would work but misses some cases the non-hir implementation currently covers.

Take this example. Here, we know that it's safe to access `a.b` on line 5 even though no dominator of its basic block accesses a.b.
```js
0: if(...) {
1:   a.b;
2: } else {
3:   a.b;
4: }
5: // It's safe to take `a.b` as a dependency of scope@0 (i.e. hoist the a.b access)
6: scope @0 {
7:   if (cond) {
8:     // access a.b;
9:   }
10:}
```

I believe a more precise formation would be that `a.b` is safe to read in $bb_i$ if any of the following hold:
- $bb_i$ itself accesses a.b
- a.b is safe to access in every pred of `a.b`
- a.b is safe to access in every successor of `a.b`

Basic block cycles can be handled by filtering / dropping the offending block (e.g. reducing `x = x && y && ...` to `x = y && ...`)

File: compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+
+export function collectHoistablePropertyLoads(
+  fn: HIRFunction,
+  usedOutsideDeclaringScope: Set<IdentifierId>
+): CollectHoistablePropertyLoadsResult {
+  const result = new TemporariesSideMap();
+
+  const functionExprRvals = fn.env.config.enableTreatFunctionDepsAsConditional
+    ? collectFunctionExpressionRValues(fn)
+    : new Set<IdentifierId>();
```
[the remark this answers is not in the document]
Comment (mofeiZ) [reply compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:30]: Sure, happy to add this in a later PR in the stack!

File: compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+    if (block.terminal.kind === "throw" || block.terminal.kind === "return") {
+      terminalPreds.add(blockId);
+    }
+  }
+
+  function recursivelyDeriveNonNull(
+    nodeId: BlockId,
+    kind: "succ" | "pred",
+    traversalState: Map<BlockId, "active" | "done">,
+    result: Map<BlockId, Set<PropertyLoadNode>>
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:303]: nit: name this more clearly? `nonNullablesByBlock` or similar?

File: compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+      }
+    }
+    // active neighbors can be filtered out as we're solving for the following
+    // relation.
+    // X = Intersect(X_neighbors, X)
+    // non-active neighbors with no recorded results can occur due to backedges.
+    // it's not safe to assume they can be filtered out (e.g. not intersected)
+    const neighborAccesses = Set_intersect([
+      ...(Array.from(neighbors)
+        .filter((n) => traversalState.get(n) !== "active")
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:340]: nit: it's easier to read a positive comparison, ie `traversalState.get(n) === 'done'`

File: compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+          result
+        );
+        changed ||= neighborChanged;
+      }
+    }
+    // active neighbors can be filtered out as we're solving for the following
+    // relation.
+    // X = Intersect(X_neighbors, X)
+    // non-active neighbors with no recorded results can occur due to backedges.
+    // it's not safe to assume they can be filtered out (e.g. not intersected)
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:337]: :+1:

File: compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+        path: [],
+      }
+    );
+  }
+
+  visitProperty(object: Place, property: string): void {
+    const nextDependency = this.#getProperty(object, property, false);
+    // if (object.identifier.id === 32) {
+    //   console.log(printDependency(nextDependency));
+    // }
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts:342]: :-)

File: compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts (hunk 2 above)
Comment (josephsavona) [reply compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts:284]: Gotcha. That makes sense. I suspect this can still be simplified - in particular, i had to remap the variable names in my head to help keep track of things. See comments for the most unclear cases.

One thing that strikes me about this algorithm is that it isn't accurate in the sense that it doesn't account for the actual live range of values. During the initial collectNodes stage we account for mutable ranges when deciding if a property access can locally be considered to be non-nullable. But then in the deriveNonNull phase we don't have access to the mutable ranges anymore. AFAICT, if you have:

```
bb0:
  let x = null
  if cond then bb1 else bb2
bb1:
  x = {}
  goto bb3
bb2:
  ...
  goto bb3:
bb3:
  x.foo.bar
```

this algorithm will infer `x.foo.bar` as non-nullable for bb0, via the intersection of bb1 & bb2 which in turn comes from bb3. So the results are actually inaccurate, but that's okay because of how we filter out dependencies in PropagateScopeDeps. 

This isn't necessarily a blocker, but it isn't great to have a pass that returns results that are technically incorrect — it's very easy to accidentally use them in a different way later, w/o realizing that important context. Thoughts on this, and whether we could make the results guaranteed accurate?

File: compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts [hunk 8]
```diff
@@ -0,0 +1,578 @@
...
+  scopeId: ScopeId,
+  tree: ReactiveScopeDependencyTreeHIR
+): void {
+  let nonNullObjects: Array<ReactiveScopeDependency> | null = null;
+  for (const [_blockId, node] of nodes) {
+    if (node.scope === scopeId) {
+      nonNullObjects = [...node.assumedNonNullObjects].map((n) => n.fullPath);
+      break;
+    }
+  }
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts:526]: maybe we should translate `nodes` into a Map keyed by scope earlier?

File: compiler/packages/babel-plugin-react-compiler/src/Entrypoint/Pipeline.ts [hunk 9]
```diff
@@ -306,6 +307,13 @@ function* runWithEnvironment(
     });
     assertTerminalSuccessorsExist(hir);
     assertTerminalPredsExist(hir);
+
+    propagateScopeDependenciesHIR(hir);
+    yield log({
+      kind: "hir",
+      name: "PropagateScopeDependenciesHIR",
+      value: hir,
+    });
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/Entrypoint/Pipeline.ts:316]: Given that `enableReactiveScopesInHIR` has been enabled by default for some time with strong validation of the results, i think as a precursor to this diff we should promote enableReactiveScopesInHIR to stable, removing the alternative code path. Let's repurpose the feature flag for the remainder of the conversion, including this one.

Basically, it would be nice to just compare the old/new PropagateScopeDeps behavior without also including all the other bugs from non-enableReactiveScopesInHIR mode.

**Review by josephsavona — COMMENTED** [review]
Ok i spent a bunch of time reading though. There's a ton here, but a lot of the core logic is retained from the existing implementation which makes it a bit easier. See comments. In particular, the note about updating our feature flag setup — it's time to drop the non-HIR based versions of older passes so that we can more easily compare the results of this new pass, and fall back quickly if necessary.

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/destructure-array-assignment-to-context-var.expect.md [hunk 10]
```diff
@@ -30,15 +30,15 @@ import { identity } from "shared-runtime";
 function Component(props) {
   const $ = _c(4);
   let x;
-  if ($[0] !== props.value) {
+  if ($[0] !== props) {
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/destructure-array-assignment-to-context-var.expect.md:33]: hmm this seems wrong, `props.value` is unconditionally accessed and should be retained as the dep

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/destructure-array-declaration-to-context-var.expect.md [hunk 11]
```diff
@@ -29,15 +29,15 @@ import { identity } from "shared-runtime";
 function Component(props) {
   const $ = _c(4);
   let x;
-  if ($[0] !== props.value) {
+  if ($[0] !== props) {
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/destructure-array-declaration-to-context-var.expect.md:32]: same here

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/functionexpr–conditional-access.expect.md [hunk 12]
```diff
@@ -25,11 +25,11 @@ import { c as _c } from "react/compiler-runtime"; // @enableTreatFunctionDepsAsC
 function Component(props) {
   const $ = _c(5);
   let t0;
-  if ($[0] !== props) {
+  if ($[0] !== props.bar) {
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/functionexpr–conditional-access.expect.md:28]: nice

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/iife-return-modified-later-phi.expect.md [hunk 13]
```diff
@@ -38,10 +38,11 @@ function Component(props) {
     items = t0;
 
     items?.push(props.a);
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/iife-return-modified-later-phi.expect.md:40]: nice, so we know that `props` is non-null and therefore can hoist the `props.a`, even though that property load is optional

**Review by josephsavona — CHANGES_REQUESTED** [review]
See previous comments, also from the fixtures it looks like there is an issue with the non-null inference

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/destructure-array-assignment-to-context-var.expect.md (hunk 10 above)
Comment (mofeiZ) [reply compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/destructure-array-assignment-to-context-var.expect.md:33]: Good catch! This is a bit of a weird one. The reason we're not inferring `props.value` as unconditionally accessed is because we check mutable ranges before adding to `assumedNonNullObjects`. This is to avoid recording loads within an object's mutable range (e.g. `read(x.y.z); x.y = null;`)
```js
// in function collectNodes
      if (value.kind === "PropertyLoad") {
        const propertyNode = c.declareProperty(...);
        if ( !isMutable(instr, value.object))  // <----- this line
          let curr = propertyNode.parent;
          while (curr != null) {
            assumedNonNullObjects.add(curr);
            curr = curr.parent;
          }
        }
```

The edge case here is that we're inferring a mutable range for `props` due to it being aliased into a context variable and captured / mutated within the lambda.
